Implement WebGL extensions

MortimerGoro opened the pull request:
<!-- Please describe your changes on the following line: -->

This PR provides the base for implementing WebGL extensions. It comes with the following ones already implemented and passing all related WebGL conformance tests:

- OES_texture_float
- OES_texture_float_linear
- OES_texture_half_float
- OES_texture_half_float_linear
- OES_vertex_array_object

I'll submit other extensions like compressed textures in a separate PR to ease the review process. I included the 5 extensions in this PR because it's easier to show/review how the WebGL extension base works.

To pass all OES_texture_float_xxx tests I had to add some missing format conversions in WebGLRenderingContext.rs

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16893)
<!-- Reviewable:end -->

MortimerGoro committed:
Implement WebGL extensions.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/tidy/servo_tidy/tidy.py
 * @kichjang: components/script/dom/mod.rs, components/script/dom/webidls/OESTextureHalfFloatLinear.webidl, components/script/dom/webidls/OESTextureFloatLinear.webidl, components/script/dom/webidls/OESTextureFloat.webidl, components/script/dom/webidls/OESVertexArrayObject.webidl and 21 more
 * @fitzgen: components/script/dom/mod.rs, components/script/dom/webidls/OESTextureHalfFloatLinear.webidl, components/script/dom/webidls/OESTextureFloatLinear.webidl, components/script/dom/webidls/OESTextureFloat.webidl, components/script/dom/webidls/OESVertexArrayObject.webidl and 21 more
 * @emilio: components/script/dom/webgl_extensions/ext/oesvertexarrayobject.rs, components/script/dom/webgl_extensions/manager.rs, components/script/dom/webgl_extensions/ext/oestexturehalffloat.rs, components/script/dom/webglbuffer.rs, components/script/dom/webgl_extensions/extension.rs and 10 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs:
```diff
@@ -0,0 +1,86 @@
+
+#[dom_struct]
+pub struct WebGLVertexArrayObjectOES {
+    reflector_: Reflector,
```
Shouldn't this be a `WebGLObject` instead of a reflector, to match the IDL?

MortimerGoro replied:
ouch, yes It should be added to this one because it's a different object (not the extension itself). Shouldn't missing the WebGLObject give a compilation error?

jdm replied:
We haven't figured out a good way to check that automatically.

nox replied:
https://is.gd/Xz2rxa

Ô tristesse, ô désespoir...

emilio commented on components/script/dom/webgl_extensions/manager.rs:
```diff
@@ -0,0 +1,192 @@
+
+type GLenum = u32;
+
+const DEFAULT_DISABLED_TEX_TYPES: [GLenum; 2] = [
```
May you add comments on all these to the rationale/spec?

emilio commented on components/script/dom/webgl_extensions/manager.rs:
```diff
@@ -0,0 +1,192 @@
+use super::wrapper::{WebGLExtensionWrapper, TypedWebGLExtensionWrapper};
+use webrender_traits::WebGLError;
+
+type GLenum = u32;
```
There's a `GLEnum` type in `gleam::gl` I believe.

emilio commented on components/script/dom/webgl_extensions/wrapper.rs:
```diff
@@ -0,0 +1,85 @@
+/// Trait used internally by WebGLExtensionManager to store and
+// handle the different WebGL extensions in a common list.
+pub trait WebGLExtensionWrapper: JSTraceable + HeapSizeOf {
+    fn instance_or_init(&self,
```
needs docs, please :)

emilio commented on components/script/dom/webgl_extensions/wrapper.rs:
```diff
@@ -0,0 +1,85 @@
+    fn is_supported(&self, &WebGLExtensionManager) -> bool;
+    fn enable(&self, manager: &WebGLExtensionManager);
+    fn name(&self) -> &'static str;
+    fn as_any<'a>(&'a self) -> &'a Any;
```
is this really needed? why can't you just do `self as &Any` in the relevant places?

emilio replied:
Also, I _think_ you can elide the lifetime.

MortimerGoro replied:
I tried to use directly the Any trait instead of adding the method but I couldn't. Here is an explanation: http://stackoverflow.com/a/33687996

Yes, the lifetime declaration was unneeded.

emilio commented on components/script/dom/webglbuffer.rs:
```diff
@@ -112,6 +119,36 @@ impl WebGLBuffer {
     }
+
+    pub fn is_attached_to_vao(&self) -> bool {
+        self.vao_references.borrow().as_ref().map_or(false, |vaos| vaos.len() > 0)
```
`!is_empty()`

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -178,11 +186,13 @@ impl WebGLRenderingContext {
                 current_scissor: Cell::new((0, 0, size.width, size.height)),
-                current_clear_color: Cell::new((0.0, 0.0, 0.0, 0.0))
+                current_clear_color: Cell::new((0.0, 0.0, 0.0, 0.0)),
+                extension_manager: WebGLExtensionManager::new()
```
It feels weird to have a `Manager` class that is a member variable (I'd have expected it to be a singleton).

Perhaps calling `WebGLExtensionManager` just `WebGLExtensions` is ok? Not a big deal though.

MortimerGoro replied:
Ok, WebGLExtensions is a good name.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -296,6 +330,24 @@ impl WebGLRenderingContext {
+                                                 info.height(),
+                                                 info.data_type().unwrap_or(TexDataType::UnsignedByte));
+            }
+
```
nit: extra newline.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -329,6 +381,36 @@ impl WebGLRenderingContext {
         }
     }
 
+    // LINEAR filtering may be forbidden when using WebGL extensions (e.g. OES_TEXTURE_FLOAT_LINEAR)
```
Link/quote spec?

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -369,6 +451,13 @@ impl WebGLRenderingContext {
     }
 
+    #[inline]
+    fn luminance(r: u8, g: u8, b: u8) -> u8 {
```
Doc comments, at least describing a source for these constants (https://en.wikipedia.org/wiki/Relative_luminance may be enough)

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1019,14 +1260,21 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.14
     fn GetSupportedExtensions(&self) -> Option<Vec<DOMString>> {
-        Some(vec![])
+        self.extension_manager.init_once(||{
```
nit: space before brace, here and everywhere else.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -2885,23 +3166,32 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
     }
 
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8
+    #[allow(unsafe_code)]
```
Why is this needed? At a glance you haven't added any unsafe code.

emilio reviewed (commented):
Looks pretty good to me!

I have a few comments and requests though.

emilio commented:
Nice! Squash, then r=me

@bors-servo delegate+

bors-servo commented:
:v: @MortimerGoro can now approve this pull request

MortimerGoro commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit c4e6a75 has been approved by `emilio`

<!-- @bors-servo r=emilio c4e6a754a66416c07b371fd903829f104bbb3f66 -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16814) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 6059bf7 has been approved by `emilio`

<!-- @bors-servo r=emilio 6059bf75401116949ad0ae16ad63906c16f2ff34 -->

bors-servo commented:
:hourglass: Testing commit 6059bf75401116949ad0ae16ad63906c16f2ff34 with merge 50d9a2b9eb53bb34f665e861e19e8c1403f7cc58...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4113)

jdm commented:
```
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #249: gl.getVertexAttrib(2, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(2, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:389:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #250: gl.getVertexAttrib(3, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(3, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:390:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #252: gl.getVertexAttrib(1, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(1, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:392:1

  ▶ OK [expected CRASH] /webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:
  │ FAIL [expected PASS] WebGL test #6: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #7: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #14: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #15: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #22: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #23: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #30: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #31: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #65: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4 expected true got false
  │ FAIL [expected PASS] WebGL test #66: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1 expected true got false
  │ FAIL [expected PASS] WebGL test #69: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4 expected true got false
  │ FAIL [expected PASS] WebGL test #70: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ runOneTest@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:453:9
  │ maybeRunTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:382:9
  └ ImageWrapper.prototype.generateData/img<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:483:7

  ▶ ERROR [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-float-with-canvas.html
  └   → ctx.fillText is not a function

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-float-with-canvas.html:
  └ NOTRUN [expected PASS] Overall test

  ▶ ERROR [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-canvas.html
  └   → ctx.fillText is not a function

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-canvas.html:
  └ NOTRUN [expected PASS] Overall test

  ▶ TIMEOUT [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-video.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ DESCRIPTION: Verify texImage2D and texSubImage2D code paths taking video elements (RGBA/HALF_FLOAT_OES)
  │ PASS Successfully enabled OES_texture_half_float extension
  │ 
  │ 
  │ testing: video/mp4; codecs="avc1.42E01E, mp4a.40.2"
  └ testing: video/mp4; codecs="avc1.42E01E, mp4a.40.2"

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-video.html:
  └ NOTRUN [expected PASS] Overall test
```

MortimerGoro commented:
Fixed issues:
- Update expectations for `ctx.fillText is not a function` in related tests
- Update test expectations for tex-image-with-format-and-type.html. Previously it was exiting early  because of missing format conversions.
- Fixed failing VERTEX_ATTRIB_ARRAY_BUFFER_BINDING tests (https://github.com/servo/servo/pull/16893/files#diff-7637bdaee356e2d89fec3a68089d9175R1886)

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 34474af7437b874773480e25cf59d59bd9d51be0 with merge 4d78aaaf77d28919f04985a5d8a3c27278b9d808...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4125)

MortimerGoro commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 32e23c4 has been approved by `emilio`

<!-- @bors-servo r=emilio 32e23c4db4a80f8ebe01bead141c5ca04bc6b215 -->

bors-servo commented:
:hourglass: Testing commit 32e23c4db4a80f8ebe01bead141c5ca04bc6b215 with merge b0976566fb9c79f7dc67b2ac808eb50ef4ad653f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7032), [arm32](http://build.servo.org/builders/arm32/builds/7242), [arm64](http://build.servo.org/builders/arm64/builds/7281), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7379), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4127), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4140), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7356), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7397), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4056), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4132), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2591)
Approved by: emilio
Pushing b0976566fb9c79f7dc67b2ac808eb50ef4ad653f to master...
